### ISSUE TITLE
Variants: Underscores or Dashes?

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#2701]
We agree that variants should be all lower-case.  If package authors want to make multi-word variant names, should they use underscores (eg: `build-type`) or dashes (eg: `build_type`)?  Note that we already use dashes for package names.

I'm voting for dashes because that's already used for package names.

**Comment by davydden** [comment]
> I'm voting for dashes because that's already used for package names.

:+1:

**Comment by citibeth** [comment]
I see, that is ambiguous.

@tgamblin Can we stop using `-` as the opposite of `+`?  We already have
`~` for that purpose, using `-` also just serves to confuse things.

**Comment by adamjstewart** [comment]
I like the minus because it's the opposite of plus 😃

**Comment by citibeth** [comment]
Well... I'm not seeing anything I like at this point.  A few principles/constraints have been brought forth, explicitly or implicitly.  They are all good, but the are not all achievable together:

1. Packages and variants should have the same naming rules.

2. The rules for specs are (after a lot of thinking about it): "all lower-case, dashes-no-dots."  Dashes are considered more user-friendly than underscores (according to LISP afficianados).

3. `-` is the opposite of `+` in specifying variants.

4. Spaces are optional in the spec.

5. Syntax cannot be ambiguous.  This is a hard constraint, it must be satisfied.

6. Bias toward how things are currently.

I see the following possible solutions:

1. Choose underscores for variants and give up on (1)

2. Get rid of `-` as a variant specifier and compromise on (5) and (3).

3. Choose underscores for packages and variants.  With that choice we get (1), but we compromise on (2) and (5).

I still think our best approach is Solution 2.  But I'm open to other possible solutions I haven't identified, or arguments why something else is better, after taking all factors into account.

**Comment by becker33** [comment]
The problem with removing `-` for specifying variants as false is that `~variant` will fail if the variant names happens to be identical to a username on the system. We could require users to use the `variant=false` syntax in that case, but that is less than ideal.

**Comment by tgamblin** [comment]
I agree with @citibeth's whole comment.  Those are the tradeoffs.  There are a couple of other ones that should probably be on the list for completeness. (or at least so I remember next time I look at this):

1. Make the parser more context-aware and parse only known package and variant names.  This sounds nice until you realize that it means a spec from the past may parse differently from a spec now.  It also makes it hard to parse specs for imaginary packages, something we can do now.  So this has all sorts of issues.
1. Compromise on (4) above and require spaces before variants in specs on the command line.

@becker33:
> The problem with removing - for specifying variants as false is that ~variant will fail if the variant names happens to be identical to a username on the system.

This is true.  However the user *can* always quote the string, so I'm not completely opposed to getting rid of `-`.  I do like it that you don't have to quote to specify variants, though.

I'm torn between two options:

1. @citibeth's (1) is easiest because it requires no additional implementation -- we just need to ask people to use `_`.  A check would be good if we pick it, though.
2. I also like my (2) above... which would amount to "go with dashes and require spaces before negative variants in a Spec.  IMHO this wouldn't be the worst thing.  We actually already require spaces before key/value options, e.g. in `spack install foo os=frontend`.

Thoughts?

**Comment by citibeth** [comment]
I like (2).  Would we then get rid of `~`?

I'd also be happy with getting rid of `-` to mean negative.  @becker33's problem only happens when users have an unlucky name, AND they want to use a space before the variant.  That can always be avoided by avoiding the space.  For example, `mypackage~myname` instead of `mypackage ~myname` (try it out in Bash and see).

Whatever we choose... I think we should consider getting rid of variations that complicate our life but don't add functionality.  Specifically:
 1. Whether or not we have a space before variants
 2. Whether we use `-` or `~`.

I think that having ONE way to do both of those things is more important that the option we choose.  Then the underscore-vs-dash problem will become easier.

**Comment by adamjstewart** [comment]
I vote for disallowing dashes in variant names. It would be easy to add a check, and Spack would continue to "just work" however you want it to.

**Comment by citibeth** [comment]
Unless somebody wants to put time into something more ambitious, I think we
should go with @adamjstewart's proposal.  If we REALLY don't like it, we
could  convert later (probably as part of a general syntax overhaul).

**Comment by tgamblin** [comment]
works for me for now.

**Comment by tgamblin** [comment]
Since @becker33 is rewriting the parser, what do people think of disallowing dashes entirely for `v1.0`?  This would eliminate confusion and there would only be one sigil for negative variants.    We would then allow `+foo`, `~foo`, `foo=true`, and `foo=false`.  Currently `-foo` is allowed, but *only* if preceded by a space.

**Comment by citibeth** [comment]
@becker33 says he will require a space, which I'm fine with.

I would lean toward `-foo`, but disallowing `~foo`, since that solves the
problem of the user with the unlucky name.  As long as a space is required
between variants, any non-whitespace character will be legal in variant
names; except that the first character can't be `+`,`-` or `=`.  I'm fine
with that.

**Comment by becker33** [comment]
My rewrite would not require a space before `+foo` or `~foo`, only before and after `foo=true`.

**Comment by citibeth** [comment]
Oh... well, I'd be in favor of requiring a space everywhere.  It's more
consistent and simplifies other parsing issues.

**Comment by tgamblin** [comment]
I believe we settled on underscores, if necessary.  I prefer short and sweet if possible.
